Explorer/Assets/DCL/Multiplayer/Connections/Archipelago/Rooms/RealmRoomsProvider.cs

```diff
@@ -0,0 +1,86 @@
+    public class RealmRoomsProvider : IRealmRoomsProvider
+    {
+        private readonly IWeb3IdentityCache identityCache;
+        private readonly ICharacterObject characterObject;
+        private readonly IWebRequestController webRequestController;
+        private readonly IArchipelagoSignFlow archipelagoSignFlow;
+        private readonly ICurrentAdapterAddress currentAdapterAddress;
+
+        /// <summary>
+        /// Maintain a single instance to be able to reconfigure and share with all consumers
+        /// </summary>
+        private readonly InteriorRoom sharedRoom = new ();
+
+        private IRealmRoomStrategy? currentStrategy;
+
+        public RealmRoomsProvider(
+            IWeb3IdentityCache identityCache,
+            ICharacterObject characterObject,
+            IWebRequestController webRequestController,
+            IArchipelagoSignFlow archipelagoSignFlow,
+            ICurrentAdapterAddress currentAdapterAddress)
+        {
+            this.identityCache = identityCache;
+            this.characterObject = characterObject;
+            this.webRequestController = webRequestController;
+            this.archipelagoSignFlow = archipelagoSignFlow;
+            this.currentAdapterAddress = currentAdapterAddress;
+        }
+
+        public IRoom Room() =>
+            sharedRoom;
+
+        public IConnectiveRoom.State CurrentState() =>
+            currentStrategy?.ConnectiveRoom.CurrentState() ?? IConnectiveRoom.State.Stopped;
+
+        /// <summary>
+        /// Should be called from RealmController
+        /// </summary>
+        /// <param name="ct"></param>
+        /// <exception cref="InvalidOperationException"></exception>
+        public async UniTask StartAsync(CancellationToken ct)
+        {
+            if (currentStrategy != null && currentStrategy.ConnectiveRoom.CurrentState() is not IConnectiveRoom.State.Stopped)
+                throw new InvalidOperationException("First stop previous room before starting a new one");
+
+            string adapterUrl = await currentAdapterAddress.AdapterUrlAsync(ct);
+
+            if (adapterUrl.Contains("wss://"))
+                currentStrategy = CreateFixedConnectionRoomStrategy(adapterUrl);
+            else if (adapterUrl.Contains("https://"))
+                currentStrategy = CreateArchipelagoIslandRoomStrategy(adapterUrl);
+            else
+                throw new InvalidOperationException($"Cannot determine the protocol from the about url: {adapterUrl}");
+
+            currentStrategy.ConnectiveRoom.Start();
+        }
+
+        public UniTask StopAsync(CancellationToken ct) =>
+            currentStrategy?.ConnectiveRoom.StopAsync(ct) ?? UniTask.CompletedTask;
+
+        private FixedConnectionRoomStrategy CreateFixedConnectionRoomStrategy(string adapterUrl) =>
+            new (sharedRoom, webRequestController, adapterUrl);
+
+        private ArchipelagoIslandRoomStrategy CreateArchipelagoIslandRoomStrategy(string adapterUrl) =>
```

Comment:
I remember points from you about that this object is factory, why it has hardcoded dependencies and acts as Factory, but named as strategy?

Reply:
Yup, `Factory` should not contain a reference to the produced object. But I agree. Possibly it should be a factory, and then it should capture a `URL` argument to produce the object. I will change it to see if it is better

Reply:
I tried to create a factory by book from it but it would require capturing variables (`IConnectiveRoom`) and I don't see any sense of doing it when we can keep them in the class. 

I don't want to introduce any other layers just because it does not fit the name of `Factory` or `Strategy`. If you know a commonly accepted name for a type that constructs an entity and holds a reference to it, I am all ears.

Reply:
We can create a IRealmRoomFactory, and inject it inside the object

I don't think merging 2 concerns Factory/Strategy into a single object is a good Idea

Reply:
I am for simplicity: I don't see the reason for 2 layers here. It's simply not needed